homeassistant/core.py

```diff
@@ -1363,7 +1363,7 @@ def is_allowed_external_url(self, url: str) -> bool:
         return any(
             allowed
             for allowed in self.allowlist_external_urls
-            if url.startswith(allowed)
+            if str(yarl.URL(url)).startswith(allowed)
```

Comment:
Now we're going to normalize the url inside every iteration of the loop. You should compute it before the loop.

Reply:
Also realize now that `startswith` is dangerous, because now `http://example.com/bla` would also allow `http://example.com/bla_something`. I think that we should make sure when we load the external urls that they all end with a `/`

Reply:
If we automatically stick a `/` onto the end of every URL, that will kill instances like `http://a.com/image.jpg` (`http://a.com/image.jpg/` won't resolve).

Reply:
The urls that we allow, not the url we're testing

Reply:
I'm wondering if we can use YARL when loading allowed URLs and then do some more intelligent comparison of both host and path. Will chew on it.

Reply:
You're still going to have the same problem if someone explicitly provides `http://a.com/image.jpg` as an allowed URL; when that is loaded from config, it will be stored as `http://a.com/image.jpg/` (against which `http://a.com/image.jpg` won't pass).

Reply:
Ah yes. In that case we can test for exact match or starts with `url + "/"`

Reply:
(e51a7102ee35 is my take on ^^^)